core/src/main/java/com/graphhopper/routing/AbstractBidirCHAlgo.java

```diff
@@ -0,0 +1,308 @@
+        levelEdgeFilter = tmpFilter;
+    }
+
+    @Override
+    public boolean finished() {
+        // we need to finish BOTH searches for CH!
+        if (finishedFrom && finishedTo)
+            return true;
+
+        // changed also the final finish condition for CH
+        return currFrom.weight >= bestWeight && currTo.weight >= bestWeight;
+    }
+
+    @Override
+    boolean fillEdgesFrom() {
+        if (pqOpenSetFrom.isEmpty()) {
+            return false;
+        }
+        currFrom = pqOpenSetFrom.poll();
+        visitedCountFrom++;
+        if (fromEntryCanBeSkipped()) {
+            return true;
+        }
+        if (fwdSearchCanBeStopped()) {
+            return false;
+        }
+        bestWeightMapOther = bestWeightMapTo;
+        fillEdges(currFrom, pqOpenSetFrom, bestWeightMapFrom, outEdgeExplorer, false);
+        return true;
+    }
+
+    @Override
+    boolean fillEdgesTo() {
+        if (pqOpenSetTo.isEmpty()) {
+            return false;
+        }
+        currTo = pqOpenSetTo.poll();
+        visitedCountTo++;
+        if (toEntryCanBeSkipped()) {
+            return true;
+        }
+        if (bwdSearchCanBeStopped()) {
+            return false;
+        }
+        bestWeightMapOther = bestWeightMapFrom;
+        fillEdges(currTo, pqOpenSetTo, bestWeightMapTo, inEdgeExplorer, true);
+        return true;
+    }
+
+    private void fillEdges(SPTEntry currEdge, PriorityQueue<SPTEntry> prioQueue,
+                           IntObjectMap<SPTEntry> bestWeightMap, RoutingCHEdgeExplorer explorer, boolean reverse) {
+        RoutingCHEdgeIterator iter = explorer.setBaseNode(currEdge.adjNode);
+        while (iter.next()) {
+            if (!accept(iter, currEdge, reverse))
+                continue;
+
+            final double weight = calcWeight(iter, currEdge, reverse);
+            if (Double.isInfinite(weight)) {
+                continue;
+            }
+            final int origEdgeId = getOrigEdgeId(iter, reverse);
+            final int traversalId = getTraversalId(iter, origEdgeId, reverse);
+            SPTEntry entry = bestWeightMap.get(traversalId);
+            if (entry == null) {
+                entry = createEntry(iter, origEdgeId, weight, currEdge, reverse);
+                bestWeightMap.put(traversalId, entry);
+                prioQueue.add(entry);
+            } else if (entry.getWeightOfVisitedPath() > weight) {
+                prioQueue.remove(entry);
+                updateEntry(entry, iter, origEdgeId, weight, currEdge, reverse);
+                prioQueue.add(entry);
+            } else
+                continue;
+
+            if (updateBestPath) {
+                // only needed for edge-based -> skip the calculation and use dummy value otherwise
+                double edgeWeight = traversalMode.isEdgeBased() ? calcWeight(iter, reverse, NO_EDGE) : Double.POSITIVE_INFINITY;
+                updateBestPath(edgeWeight, entry, origEdgeId, traversalId, reverse);
```

Comment:
Here the calcWeight operation is done although it could happen that the backward search has no entry and the weight is not used. Might be related to (not very likely):

> So far the query speed seems to decrease by about 10%, I could not find out why exactly this is, 

BTW: does this slowdown occur for edge- and node-based queries?

Reply:
oh good point, its even worse: edge-based CH overrides the updateBestPath method and does not need edgeWeight in any case. will try if this makes the difference but I would guess so.

> BTW: does this slowdown occur for edge- and node-based queries?

I think it is more pronounced for edge-based, but even after running many measurements I was not entirely sure

Reply:
No, unfortunately this did not fix it. My next best guess is that `RoutingCHGraph#getEdgeIteratorState` creates two (vs. one) objects, which is relevant for path unpacking. So this could be fixed with #1847